### DISCUSSION TITLE
How can we remove a rule from the knowledge ?

### DISCUSSION THREAD

**Discussion by atanuroy007** [evrete/evrete#32] [Q&A]
Hi,

I am trying to add/remove/modify rules in runtime and not able to make remove working.

Trying like- 

```
if(knowledge.ruleExists(id)) {
	RuleDescriptor rule  = knowledge.getRule(id);
	knowledge.getRules().remove(rule);
}
```

But this is not working, throwing below error - 

java.lang.UnsupportedOperationException
        at java.util.Collections$UnmodifiableCollection.remove(Collections.java:1071)

Please could you help?

**Comment by andbi** [comment]
First, why the exception:

In most rule engines, rules do not exist independently - they are always part of a ruleset. Engines analyze rules in their entirety to allocate session memory most efficiently, reuse memory components, and avoid duplicate conditions.

Regarding your question, if by "remove" you mean to make a rule stop activating, then there are several ways:

1. Make the ruleset itself define which rules ought to be fired. Often it can be done by including an additional "flag" object in the rule's condition block, e.g., `"$flag.enabled == true"`. This way, your logic will be consistently defined by rules rather than by a mix of rules and Java code.
2.  Use the `ActivationManager` API as described in https://evrete.org/docs/conflict-resolution/#activation-manager

**Reply by atanuroy007** [reply]
@andbi thank you very much! 
If we set ActivationMode = CONTINUOUS, then can we use knowledge.getRules().remove(rule); ?
On the first point, how can we change a rule in runtime to set the flag value ?

In my use case, I am storing the rules in DB, in case a rule is add/deleted/updated in the DB, that should reflect in the rule engine as well.

Could you kindly help me with this ?

**Reply by andbi** [reply]
Give me some context, why do you want to "remove" a rule, what's the end goal? If possible, please provide a simplified example.

And no, the `CONTINUOUS` flag won't let you to remove a rule. I must've posted a wrong link, see the `ActivationManager`, not the `ActivationMode`.

**Reply by atanuroy007** [reply]
@andbi The requirement at high level is - We have several processes, each process has multiple tasks, tasks are containing multiple data elements. Based on these there combinations, we are trying to send emails to end users who will be working on the tasks. The email template can vary based on the these process+task+data combinations. A particular combination would return a template.

Admins can configure(add/update/delete) these combinations through UI. Rules are getting saved in DB.

As of now we are reloading the rule engine once everyday during non-business hours, which shuts down the service and loads the latest rules from the database. But, this is not real-time. So, we are tying to make  real-time update to the running rule engine.

I hope, I could explain. 

Please could you help with this ?

**Reply by andbi** [reply]
Got it. 

If I wanted a real-time system and I hadn't migrated to Evrete from another RBMS, I would do the following:

Instead of dealing with POJO classes with field values, I would create the domain types with functional fields which would query real-time parameters from your backend, see this link as an example of such an approach: https://evrete.org/guides/howto/field-declaration/

That would change nothing for the rule authors because fact types and property names would remain the same, but the data would be real-time. I prefer this approach for new projects as it gives a higher abstraction level and makes life easier for rule authors.

If admins need to create new rules, then it's worth a closer look at the root cause: why do they need to do that, and whether it's possible to delegate their logic (whatever it is) to the rule engine.

If neither of the above suits your situation or is challenging to implement, then it's always possible to queue template tasks and let them wait until a new ruleset is created. Or let these templates use the previous ruleset while a new one is being built (how long does it take, by the way?)

**Reply by atanuroy007** [reply]
@andbi Thanks very much for the help, I'll certainly discuss this with team, if we can implement in that way.
Would be able to help me on another problem - As I mentioned earlier, we are developing a daily scheduler which would unload the rule engine and reload the latest rules from the database.

We are using service.shutdown(); and expecting that it would clear up everything. But is not, after executing service.shutdown(); when we are trying to declare the fields and load the rules but getting below errors respectively - 

1.	java.lang.IllegalStateException: Type name 'Application' has been already defined: {name='Application', javaType='com.sp.events.event.JSEvent'}
2.	java.lang.IllegalArgumentException: Rule '1' already exists

So, please could help me here ? How can we remove everything and initialize the rule engine as a fresh?

Thanks in advance!

**Reply by andbi** [reply]
If you shut down an instance of `KnowledgeService`, you will need to create a new one and apply rulesets from scratch:
`new KnowledgeService().newKnowledge()`
, and so on.

**Reply by andbi** [reply]
What may be happening in your case, is that your team somehow keeps the instances of `Knowledge` or rules, and it's those objects that contain references to an old service.

**Reply by atanuroy007** [reply]
@andbi Yes, actually we are hosting this in WebSphere, shutdown is not really removing the instance.
When we restart the application in WebSphere, it works.

**Reply by andbi** [reply]
That's the correct usage, you only need to shutdown the service when your app server shuts down. 

Then you just need to eliminate the inconsistency. Remember, within the same service, nothing prevents you from creating a new ruleset while the old one is still in use. Once the new ruleset is ready, then you could just swap the instances and let the old one to be garbage collected.

**Reply by andbi** [reply]
```java
public class AppContainer {
    private final KnowledgeService knowledgeService = new KnowledgeService();
    
    private Knowledge currentKnowledge;
    
    @Async
    public void createNewRulesetAsync() {
        // 1. Create e new ruleset, possibly asynchronously
        Knowledge newKnowledge = knowledgeService.newKnowledge();
        
        // 2. Fill it with the rules from your backend
        // ....
        
        // 3. Swap the instances
        this.currentKnowledge = newKnowledge;
    }

    public StatelessSession newStatelessSession() {
        return currentKnowledge.newStatelessSession();
    }
    
    @PreDestroy
    public void shutdown() {
        this.knowledgeService.shutdown();
    }
}

```

**Reply by atanuroy007** [reply]
@andbi thanks very much!
I'll try this.
